kythe/go/util/compare/compare.go

```diff
@@ -0,0 +1,272 @@
+	LT Order = -1 // lhs < rhs
+	EQ Order = 0  // lhs == rhs
+	GT Order = 1  // lhs > rhs
+)
+
+// Reverse reverses the Order: LT -> GT, GT -> LT; EQ -> EQ.
+func (o Order) Reverse() Order {
+	return Order(-1 * o)
+}
+
+func (o Order) String() string {
+	switch o {
+	case LT:
+		return "LT"
+	case GT:
+		return "GT"
+	case EQ:
+		return "EQ"
+	default:
+		return fmt.Sprintf("Order(%d)", o)
+	}
+}
+
+// Strings returns LT if s < t, EQ if s == t, or GT if s > t.
+func Strings(s, t string) Order { return Order(strings.Compare(s, t)) }
+
+// Bytes returns LT if s < t, EQ if s == t, or GT if s > t.
+func Bytes(s, t []byte) Order { return Order(bytes.Compare(s, t)) }
+
+// Ints returns LT if a < b, EQ if a == b, or GT if a > b.
+func Ints(a, b int) Order { return ToOrder(a - b) }
+
+var (
+	emptyVName = new(spb.VName)
+	emptyEntry = new(spb.Entry)
+)
+
+// Options for comparing components of *spb.VName protobuf messages.
+var (
+	ByVNameSignature = By(func(x interface{}) interface{} {
+		return x.(*spb.VName).GetSignature()
+	})
+	ByVNameCorpus = By(func(x interface{}) interface{} {
+		return x.(*spb.VName).GetCorpus()
+	})
+	ByVNameRoot = By(func(x interface{}) interface{} {
+		return x.(*spb.VName).GetRoot()
+	})
+	ByVNamePath = By(func(x interface{}) interface{} {
+		return x.(*spb.VName).GetPath()
+	})
+	ByVNameLanguage = By(func(x interface{}) interface{} {
+		return x.(*spb.VName).GetLanguage()
+	})
+)
+
+// VNames returns LT if v1 precedes v2, EQ if v1 and v2 are equal, or GT if v1
+// follows v2, in standard order.  The ordering for VNames is defined by
+// lexicographic comparison of [signature, corpus, root, path, language].
+func VNames(v1, v2 *spb.VName) Order {
+	return Seq(v1, v2,
+		ByVNameSignature, ByVNameCorpus, ByVNameRoot, ByVNamePath, ByVNameLanguage)
+}
+
+// VNamesEqual reports whether v1 and v2 are equal.
+func VNamesEqual(v1, v2 *spb.VName) bool { return VNames(v1, v2) == EQ }
+
+// Options for comparing components of *spb.Entry protobuf messages.
+var (
+	ByEntrySource = And(By(func(x interface{}) interface{} {
+		return x.(*spb.Entry).GetSource()
+	}), With(func(a, b interface{}) Order { return VNames(a.(*spb.VName), b.(*spb.VName)) }))
+	ByEntryEdgeKind = By(func(x interface{}) interface{} { return x.(*spb.Entry).GetEdgeKind() })
+	ByEntryFactName = By(func(x interface{}) interface{} { return x.(*spb.Entry).GetFactName() })
+	ByEntryTarget   = And(By(func(x interface{}) interface{} {
+		return x.(*spb.Entry).GetTarget()
+	}), With(func(a, b interface{}) Order { return VNames(a.(*spb.VName), b.(*spb.VName)) }))
+)
+
+// Entries reports whether e1 is LT, GT, or EQ to e2 in entry order, ignoring
+// fact values (if any).
+//
+// The ordering for entries is defined by lexicographic comparison of
+// [source, edge kind, fact name, target].
+func Entries(e1, e2 *spb.Entry) Order {
+	return Seq(e1, e2,
+		ByEntrySource, ByEntryEdgeKind, ByEntryFactName, ByEntryTarget)
+}
+
+// ByEntries is a min-heap of entries, ordered by Entries.
```

Comment:
What makes this a min-heap? I don't think I see the bits that guarantee min-heap properties.

Reply:
It's a weird idiom.  The type implements this interface: https://godoc.org/container/heap#Interface.  As long as you interact with the type using the accompanying Push/Pop functions, it all works out.  This is type wasn't added in this commit, it just got carried over from the previous package.